### ISSUE TITLE
Need a testing framework

### ISSUE THREAD

**Issue by cbreeden** [cbreeden/ReX#9]
There is some prior work done here for related projects:
  - KaTeX uses a selenium docker to render their content in firefox and chrome.
  - KaTeX also can compare LaTeX with their output for comparisons.
  - MathJax also uses selenium, with a Web UI it appears
  - WebKit has documents with things like: long radicals, long radicals superscript, etc..

Selenium might be best for getting an accurate representation for our rendered SVG.  I proposed that we places `Rules` in the document to be used to assist in measuring distances.  Another option would be to do something similar to what KaTeX does and compare the output with LaTeX.

**Comment by s3bk** [comment]
First of all, I don't think this can be fully automated. More of a machine assisted diffing.
The advantage of ReX is the backend independent rendering, so one could write a `TestRenderer` which produces a `type TestSet = Set<(Cursor, Scale, Glyph)>;`. This output would be compared to a previous version and pass if the sets are equal up to some _ɛ > 0_.

How the different Renderer implementations compare their output is a different story.
 - SVGRenderer ↔ Canvas could be done by rasterizing the SVG and comparing the image.
 - SVGRenderer ↔ TestRenderer could be done by parsing the SVG and creating a `TestSet` from it.

**Comment by cbreeden** [comment]
Yeah this is far more realistic now that we have the absolute position of each glyph.  Rex could even construct a "diff" by painting the glyphs a different color: one red, the other blue.  This could make visualizing the changes quite nice.  I've been using Batik lately to rasterize the glyphs (it's what rasterizes the images on the front page).
